File: conversational-engine/src/conversational_engine/scripts/ensure_db.py
from __future__ import annotations

from psycopg import connect

from conversational_engine.config.settings import get_settings


def quote_identifier(identifier: str) -> str:
    return '"' + identifier.replace('"', '""') + '"'


def admin_database_url(database_url: str) -> tuple[str, str]:
    if '://' not in database_url:
        raise ValueError('Invalid database URL')

    scheme, rest = database_url.split('://', 1)
    authority, _, tail = rest.partition('/')
    database_name, separator, query = tail.partition('?')
    if not database_name:
        raise ValueError('Database URL must include a database name')

    admin_rest = authority + '/postgres'
    if separator:
        admin_rest += '?' + query
    return f'{scheme}://{admin_rest}', database_name


def main() -> None:
    settings = get_settings()
    admin_url, database_name = admin_database_url(settings.database_url)

    if database_name == 'postgres':
        print('Database URL already points to postgres; skipping database creation')
        return

    with connect(admin_url, autocommit=True) as conn, conn.cursor() as cur:
        cur.execute('SELECT 1 FROM pg_database WHERE datname = %s', (database_name,))
        if cur.fetchone():
            print(f'Database {database_name} already exists')
            return

        print(f'Creating database {database_name}')
        cur.execute(f'CREATE DATABASE {quote_identifier(database_name)}')
        print(f'Created database {database_name}')


if __name__ == '__main__':
    main()

File: backend/src/scripts/ensure-db.ts
import { Client, type ClientConfig } from 'pg';
import { buildPgClientConfig } from '@backend/db/connection.js';

function buildClientConfig(connectionString: string): ClientConfig {
  return buildPgClientConfig(connectionString);
}

function getDatabaseUrl(): URL {
  const raw = process.env.DATABASE_URL;
  if (!raw) {
    throw new Error('Missing DATABASE_URL');
  }

  return new URL(raw);
}

function getDatabaseName(url: URL): string {
  const databaseName = url.pathname.replace(/^\/+/, '');
  if (!databaseName) {
    throw new Error('DATABASE_URL must include a database name');
  }

  return databaseName;
}

function buildAdminConnectionString(url: URL): string {
  const adminUrl = new URL(url.toString());
  adminUrl.pathname = '/postgres';
  return adminUrl.toString();
}

function quoteIdentifier(identifier: string): string {
  return `"${identifier.replace(/"/g, '""')}"`;
}

async function main() {
  const databaseUrl = getDatabaseUrl();
  const databaseName = getDatabaseName(databaseUrl);

  if (databaseName === 'postgres') {
    console.log('DATABASE_URL already points at postgres; skipping database creation');
    return;
  }

  const adminClient = new Client(buildClientConfig(buildAdminConnectionString(databaseUrl)));

  try {
    await adminClient.connect();
    const exists = await adminClient.query(
      'SELECT 1 FROM pg_database WHERE datname = $1',
      [databaseName]
    );

    if (exists.rowCount && exists.rowCount > 0) {
      console.log(`Database ${databaseName} already exists`);
      return;
    }

    console.log(`Creating database ${databaseName}`);
    await adminClient.query(`CREATE DATABASE ${quoteIdentifier(databaseName)}`);
    console.log(`Created database ${databaseName}`);
  } finally {
    await adminClient.end();
  }
}

main().catch((error) => {
  console.error('Database bootstrap failed:', error);
  process.exit(1);
});
